### DISCUSSION TITLE
Extract regexp route with hash_routes

### DISCUSSION THREAD

**Discussion by muellerj** [jeremyevans/roda#276] [General]
I would like to extract a route block from the main application file and thought that the `hash_routes` plugin would be suitable. I can't manage to get it to work though.

The route blocks in question before the extraction look like this:

``` ruby
# app.rb

class App < Roda

  # ...

  require_relative "routes1"   # See comments below
  require_relative "routes2"   # See comments below

  route do |r|
    r.hash_routes

    r.root do
      "root route"
    end

    # Extracting this is no problem
    r.on "foo" do
      r.get true do
        "foo route"
      end
    end

    # Extracting is is harder?
    r.is %r{(.*)} do |slug|
      r.pass unless @page = Page.find(slug)
      # ...
    end
  end

end
```

Extracting the `foo` route via `hash_routes.on("foo") do |r|` works as expected:

``` ruby
# routes1.rb
class App < Roda
  hash_routes.on "foo" do |r|
    r.get true do
      "foo route"
    end
  end
end
```

However, this does not match any path as expected:

``` ruby
# routes2.rb
class App < Roda
  hash_routes do |r|
     r.is %r{(.*)} do |slug|
       r.pass unless @page = Page.find(slug)
       # ...
    end
  end
end
```

Is there a trick I'm missing? Thanks in advance for your help!

**Comment by jeremyevans** [comment]
You cannot use regexps when routing in hash routes.  The arguments you provide when routing in hash routes need to be strings.

`%r{(.*)}` looks like it would match almost anything.  Assuming you do want to match anything, you might be able to replace it with:

```ruby
  route do |r|
    r.hash_routes

    r.root do
      "root route"
    end

    # Maybe r.path_info[1..-1] if you don't want the starting slash
    if @page = Page.find(r.path_info)
      # ...
    end
  end
```

**Reply by muellerj** [reply]
Aye, that was the idea - match anything and see if there's a page that matches the slug.

The block you're suggesting would still have to live in the main `route do |r|` block?

**Reply by jeremyevans** [reply]
In my example, it is in the main route block, but that's not required. You could use the `named_routes` plugin if you want to move it outside of the main route block.

**Reply by muellerj** [reply]
@jeremyevans: As ususal thanks a lot for your prompt help!

If someone stumbles across this, this is what I ended up doing:

``` ruby
# app.rb

class App < Roda

  # ...
  plugin :multi_route

  require_relative "routes1"   # See comments below
  require_relative "routes2"   # See comments below

  route do |r|
    r.hash_routes

    r.root do
      "root route"
    end

    r.route "pages"
  end

end
```

``` ruby
# routes1.rb
class App < Roda
  hash_routes.on "foo" do |r|
    r.get true do
      "foo route"
    end
  end
end
```

``` ruby
# routes2.rb
class App < Roda
  route "pages" do |r|
    r.is %r{(.*)} do |slug|
      r.pass unless @page = Page.find(slug)
      
      # ...

    end
  end
end
```
